### ISSUE TITLE
mixed_precision returns gradient zeros when the model input size is large

### ISSUE THREAD

**Issue by WingsOfPanda** [tensorflow/tensorflow#52236]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Noe
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.6.0-rc2-32-g919f693420e 2.6.0
- Python version: 3.9
- CUDA/cuDNN version: 11.1/8.1.1
- GPU model and memory: RTX 3090/24GB

**Describe the current behavior**
When using mixed_precision policy described in [https://www.tensorflow.org/guide/mixed_precision](url) with large model input size, for example `(256, 368, 368,)`, the returned gradient are constantly ZEROS. However, if remove the mixed_precision policy, the returned gradient is normal with non-zeros numbers.

Moreover, if we use small model input size, let's say `(16, 16, 16)`, the returned gradient is normal no matter the mixed_precision is allowed or not. 

My model is a typical U-net like model.

**Describe the expected behavior**
With large model input size like `(256, 368, 368)` used above, the returned gradient should be at least non-zeros. Otherwise the model won't be trained. 

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

The code below set the input shape to `[1, 256, 368, 368, 1]` and allows `mixed_precision.Policy('mixed_float16')`. It will return zeros gradients in the end (hence, no training at all). Setting `tf16_flag=False` will returns normal gradient behavior.

Also, by change `shape = [1, 16, 16, 16, 1]`, the gradient behaves normally no matter allows `mixed_precision.Policy('mixed_float16')` or not

```
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras import mixed_precision
import numpy as np
from tqdm import tqdm

gpus = tf.config.experimental.list_physical_devices('GPU')

for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)

tf16_flag = True

if tf16_flag:
    policy = mixed_precision.Policy('mixed_float16')
    mixed_precision.set_global_policy(policy)


shape = [1, 256, 368, 368, 1]
# shape = [1, 16, 16, 16, 1]


def forward_conv(x, filters, kernels, name='forward', padding='same'):
    i = 0
    for flt, kernel in zip(filters, kernels):
        x = layers.Conv3D(flt, kernel, activation='relu', padding=padding, dilation_rate=(1, 1, 1),
                          use_bias=False, name=str(i) + '_' + name)(x)
        x = layers.BatchNormalization(name=str(i) + '_bn_' + name)(x)
        i += 1
    return x


def part_one(ipt):
    l1 = forward_conv(ipt, (4, 4), (3, 3), name='enc1')
    d2 = layers.MaxPool3D(pool_size=(2, 2, 2))(l1)
    l2 = forward_conv(d2, (4, 4), (3, 3), name='enc2')
    return l1, l2


def part_inner(ipt1, ipt2):
    l1 = forward_conv(ipt1, (4, 4), (3, 3), name='enc1')
    l2 = forward_conv(ipt2, (4, 4), (3, 3), name='enc2')
    return l1, l2


def part_two(ipt1, ipt2):
    l2 = forward_conv(ipt2, (4, 4), (3, 3), name='dec2')
    u1 = layers.UpSampling3D(size=(2, 2, 2))(l2)
    r1 = forward_conv(ipt1 + u1, (4, 4), (3, 3), name='dec1')
    return r1

initial = tf.ones(shape, dtype=tf.float16) if tf16_flag \
    else tf.ones(shape, dtype=tf.float32)

tf.random.set_seed(1)

with tf.GradientTape() as g:
    g.watch(initial)
    l1_, l2_ = part_one(initial)
    for _ in range(2):
        l1_, l2_ = part_inner(l1_, l2_)
    opt_ = part_two(l1_, l2_)
    loss = tf.reduce_mean(l1_) + tf.reduce_mean(opt_)
    gd = g.gradient(loss, initial)
    print('-' * 100)
    print(f'loss is {loss} and grad is {np.sum(gd)} with input shape {shape}')

```

**Comment by tilakrayal** [comment]
@sanatmpa1 ,
I was able to reproduce the issue in tf v2.4 and v2.6 where as in nightly version the code is executed with different error.Please find the gist [here](https://colab.research.google.com/gist/tilakrayal/64331ba9914ff5c473afdfe7924dc4a1/52236.ipynb).

**Comment by WingsOfPanda** [comment]
@tilakrayal @sanatmpa1 thx for your response! I think you need to have at least 24GB gpu to run this code with input size `(256, 368, 368)' otherwise OOM will happen.

**Comment by sachinprasadhs** [comment]
Since this is a Keras issue, Could you please move this to https://github.com/keras-team/keras.

**Comment by WingsOfPanda** [comment]
@sachinprasadhs since this issue can be reproduced in tf2.4. I think it is still part of tf?

**Comment by sachinprasadhs** [comment]
@WingsOfPanda , Yes, since `mixed_precision` is under  `tensorflow.keras`  it would be good to move this issue under https://github.com/keras-team/keras so that we can track keras issues separately under the mentioned repository.
